rpc-rs/src/rpc_client.rs

```diff
@@ -208,21 +256,86 @@ impl RpcClient {
+        let len = message.arg.len();
+        let chunkify = {
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "chunkify")] {
+                    crate::chunkify::needs_chunking(len)
+                } else {
+                    false
+                }
+            }
+        };
+
+        #[allow(unused_variables)]
+        let (invocation, body) = {
+            let mut inv = Invocation {
+                origin,
+                target,
+                operation: method.clone(),
+                id: subject,
+                encoded_claims: claims.encode(&self.key).unwrap(),
+                host_id: self.host_id.clone(),
+                content_length: Some(len as u64),
+                ..Default::default()
+            };
+            if chunkify {
+                (inv, Some(Vec::from(message.arg)))
+            } else {
+                inv.msg = Vec::from(message.arg);
+                (inv, None)
+            }
+        };
+        let nats_body = crate::common::serialize(&invocation)?;
+
+        let chunkify_join: Option<JoinHandle<RpcResult<()>>> = {
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "chunkify")] {
+            if let Some(body) = body
+            {
+                let inv_id = invocation.id.clone();
+                debug!("chunkifying inv {} size {}", &inv_id, len);
+                // start chunking thread
+                let lattice = self.lattice_prefix().to_string();
+                Some(tokio::task::spawn_blocking(move || {
+                    let ce = chunkify_endpoint(None, lattice)?;
+                    ce.chunkify(&inv_id, &mut body.as_slice())
+                }))
+            } else {
+                None
+            }
+
+                    } else {
+                       None
+                    }
+                }
+        };
+        // I tried starting the send to ObjectStore in background thread,
+        // and then send the rpc, but if the objectstore hasn't completed,
+        // the recipient gets a missing object error,
+        // so we need to flush this first
+        if let Some(join) = chunkify_join {
+            let _ = join.await.map_err(|e| RpcError::Other(e.to_string()))?;
+            debug!("writing to ObjectStore should be complete");
+        }
+
+        let timeout = if chunkify {
+            timeout.map(|t| t + CHUNK_RPC_EXTRA_TIME)
+        } else {
+            timeout
         };
         trace!("rpc send {}", &target_url);
 
-        let nats_body = crate::serialize(&invocation)?;
         if expect_response {
             let payload = if let Some(timeout) = timeout {
-                match tokio::time::timeout(timeout, self.request(&topic, &nats_body)).await {
+                // if we expect a response before timeout, finish sending all chunks, then wait for response
+                match tokio::time::timeout(
+                    timeout,
+                    //self.request_join(&topic, &nats_body, chunkify_join),
```

Comment:
we should remove the comments from experiments and research to avoid confusion

Reply:
done